File: cpp/cppfind/src/FileResult.cpp
#include "FileResult.h"

namespace cppfind {
    FileResult::FileResult(const std::filesystem::path& file_path, const FileType file_type, const uint64_t file_size,
        const long last_mod) :
    m_file_path{file_path}, m_file_type{file_type}, m_file_size{file_size}, m_last_mod{last_mod} {
        this->m_containers = {};
    }

    FileResult::FileResult(const std::vector<std::filesystem::path>& containers, const std::filesystem::path& file_path,
                           const FileType file_type, const uint64_t file_size, const long last_mod) :
    m_containers{containers}, m_file_path{file_path}, m_file_type{file_type}, m_file_size{file_size},
    m_last_mod{last_mod} {
    }

    std::vector<std::filesystem::path> FileResult::containers() const {
        return m_containers;
    }

    std::filesystem::path FileResult::file_path() const {
        return m_file_path;
    }

    std::string FileResult::file_name() const {
        return m_file_path.filename().string();
    }

    FileType FileResult::file_type() const {
        return m_file_type;
    }

    uint64_t FileResult::file_size() const {
        return m_file_size;
    }

    long FileResult::last_mod() const {
        return m_last_mod;
    }

    std::string FileResult::string() const {
        std::string fullpath{};
        for (const auto& c : m_containers) {
            fullpath.append(c).append(CONTAINER_SEPARATOR);
        }
        fullpath.append(m_file_path.string());
        return fullpath;
    }
}

File: php/phpfind/src/phpfind/FileResult.php
<?php

declare(strict_types=1);

namespace phpfind;

/**
 * Class FileResult
 *
 * @package phpfind
 * @property string[] $containers
 * @property string $path
 * @property string $file_name
 * @property int $file_size
 * @property int $last_mod
 */

class FileResult
{
    const string CONTAINER_SEPARATOR = '!';

    /**
     * @var string[] $containers
     */
    public array $containers;
    public readonly string $file_path;
    public readonly FileType $file_type;
    public readonly int $file_size;
    public readonly int $last_mod;

    /**
     * @param string $file_path
     * @param FileType $file_type
     * @param int $file_size
     * @param int $last_mod
     */
    public function __construct(string $file_path, FileType $file_type, int $file_size, int $last_mod)
    {
        $this->containers = [];
        $this->file_path = $file_path;
        $this->file_type = $file_type;
        $this->file_size = $file_size;
        $this->last_mod = $last_mod;
    }

    /**
     * @return string
     */
    public function path(): string
    {
        return dirname($this->file_path);
    }

    /**
     * @return string
     */
    public function file_name(): string
    {
        return basename($this->file_path);
    }

    /**
     * @return string
     */
    public function __toString(): string
    {
        $s = "";
        if ($this->containers) {
            $s = join(FileResult::CONTAINER_SEPARATOR, $this->containers) .
                FileResult::CONTAINER_SEPARATOR;
        }
        $s .= FileUtil::join_paths($this->path, $this->file_name);
        return $s;
    }
}
